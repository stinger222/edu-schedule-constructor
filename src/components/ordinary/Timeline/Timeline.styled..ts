import styled from "styled-components";

export const StyledTimeline = styled.div`
	display: flex;
	width: 6.5em;

	& .label {
		height: 100%;
		width: 75%;
		font-size: 1.5em;
		margin-right: 0.4em;
	}
	
	& .label > div {
		text-align: center;
		font-weight: 600;
	}

	& .label .label-end {
		font-size: 0.85em;
		color: ${({theme}) => theme.text.secondary};
	}

	& .progress-bar-body {
		height: 100%;
		width: 25%;
	}

	& .progress-bar-body .indicator {
		width: 1em;
		height: 1em;
		background: transparent;
		box-shadow: 0 0 0 0.15em #0075FF inset;
		border-radius: 5em;
		margin-inline: auto;
		margin-bottom: 0.4em;
	}

	&.active .progress-bar-body .indicator {
		height: 1.3em;
		width: 1.3em;
		position: relative;
		box-shadow: 0 0 0 0.13em #0075FF inset;
	}

	&.active .progress-bar-body .indicator:after {
		content: "";
		position: absolute;
		inset: 50%;
		transform: translate(-50%, -50%);
		height: 0.4em;
		width: 0.4em;
		box-shadow: 0 0 0 0.113em #0075FF;
		/* box-shadow: 0 0 0 0.2em red; */
		border-radius: 10em;
		background: #0075ff;
	}
	
	& .progress-bar-body .line {
		width: 0.2em;
		height: calc(100% - 0.8em);
		border-radius: 1em;
		background: #0075FF;
		margin-inline: auto;
	}
`
