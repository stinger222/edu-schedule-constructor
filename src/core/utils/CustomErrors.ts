export class EmptyDay extends Error {
  readonly name = "EmptyDay"

  constructor() {
    super()
  }
}

export class Warning extends Error {
  readonly name = "Warning"

  constructor(message: string) {
    super(message)
  }
}
