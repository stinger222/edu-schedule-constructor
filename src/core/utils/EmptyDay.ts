class EmptyDay extends Error {
  name = "EmptyDay"

  constructor(message: string) {
    super(message)
  }
}

export default EmptyDay
