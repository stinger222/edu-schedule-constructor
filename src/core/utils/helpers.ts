import { weekDates, Replacements, Cases } from './../types/types';

export const getCurrentWeekDates = (): weekDates => {
	const result = []
	const date = new Date()
	date.setDate(date.getDate() - (date.getDay() === 0 ? 7 : date.getDay())) // now it's current week sunday date
	for (let i = 0; i < 7; i++) {
		date.setDate(date.getDate() + 1) // and now it's monday
		result.push(date.getDate())
	}

	return formatNumbers(result) as weekDates
}

export const formatNumbers = (numbers: number[]): string[] => {
	return numbers.map(number => formatNumber(number))
}

export function replaceBlankProps<P extends {[key: string]: any}> (props: P , replacements: Replacements<P>): P {
	const result: any = {}
	for (const [key, value] of Object.entries(props)) {
		if (typeof value === 'string') {
			if (value.trim() == '') {
				result[key] = replacements[key as keyof typeof replacements]
			} else {
				result[key] = value
			}
		} else {
			result[key] = value
		}
	}
	return result as P
}

export const formatNumber = (number: number): string => {
	if (isNaN(number)) return "??"
	
	return Math.abs(number).toLocaleString('en-US', {
    minimumIntegerDigits: 2,
    useGrouping: false
  })
}

export const formatTimeString = (timeString: string): string => {
	let [hours, minutes] = timeString.split(':')

	return `${formatNumber(parseInt(hours))}:${formatNumber(parseInt(minutes))}`
}

export class WeekDays {
	private static _week = {
		short: ['Пн', 'Вт', 'Ср', 'Чт', 'Пт', 'Сб', 'Вс'],
		full: {
			nominative: [
				'Понедельник', 'Вторник', 'Среда', 'Четверг', 'Пятница', 'Суббота', 'Воскресенье'
			],
			accusative: [
				'Понедельник', 'Вторник', 'Среду', 'Четверг', 'Пятницу', 'Субботу', 'Воскресенье'
			]
		} as Record<Cases, string[]>
	}

	static getShort() {
		return this._week.short
	}

	static getFull(_case: Cases = Cases.Nominative, lowerCase: boolean = false) {
		let result = this._week.full[_case]
		return lowerCase ? result.map(i => i.toLowerCase()) : result
	}
}