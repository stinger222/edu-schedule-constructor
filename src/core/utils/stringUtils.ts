export const formatNumbers = (numbers: number[]): string[] => {
	return numbers.map(number => formatNumber(number))
}

export const formatNumber = (number: number, digits: number = 2): string => {
	if (isNaN(number)) return "??"
  
	const formatter = new Intl.NumberFormat("en-US", {
		minimumIntegerDigits: digits,
		useGrouping: true
	})
  
	return formatter.format(Math.abs(number))
}

export const formatTimeString = (timeString: string): string => {
	const [hours, minutes] = timeString.split(":")

	return `${formatNumber(parseInt(hours))}:${formatNumber(parseInt(minutes))}`
}

export const capitalize = (str: string, eachWord: boolean = false): string => {
	if (eachWord) {
		return str
			.split(" ")
			.map(s => s.charAt(0).toUpperCase() + s.slice(1))
			.join(" ")
	}
	return str.charAt(0).toUpperCase() + str.slice(1)
}

export const validateField = (value: string) => {
	value = value.trim()
	return value !== "" && value !== "undefined"
}