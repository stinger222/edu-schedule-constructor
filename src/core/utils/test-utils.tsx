import React, { ReactElement } from "react"
import { HashRouter } from "react-router-dom"
import { render, RenderOptions } from "@testing-library/react"

import RootStore from "../store/RootStore"


import { ThemeProvider } from "styled-components"
import { DarkTheme } from "../themes/Dark"

const AllTheProviders: React.FC<{children: ReactElement}> = ({children}) => {

	const rootStore = new RootStore()
	const StoreContext = React.createContext<typeof rootStore>(rootStore)

	return (
		<ThemeProvider theme={DarkTheme}>
			<StoreContext.Provider value={rootStore}>
				<HashRouter>
					{children}
				</HashRouter>
			</StoreContext.Provider>
		</ThemeProvider>
	)
}

const customRender = (ui: React.ReactElement, options?: Omit<RenderOptions, "queries">) => {
	return render(ui, { wrapper: AllTheProviders, ...options})
}

export * from "@testing-library/react"
export {customRender as render}