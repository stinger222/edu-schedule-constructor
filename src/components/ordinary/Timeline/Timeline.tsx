import React from "react"
import { StyledTimeline } from "./Timeline.styled."
import { Replacements } from "../../../core/types/types"
import { formatTimeString, replaceBlankProps } from "../../../core/utils/helpers"

interface IProps {
	startTime: string,
	endTime: string,
	active?: boolean
}

const propsReplacements: Replacements<IProps> = {
	startTime: "??:??",
	endTime: "??:??",
}

const Timeline: React.FC<IProps> = (props) => {
	let { startTime, endTime, active } = replaceBlankProps<IProps>(props, propsReplacements)

	startTime = formatTimeString(startTime)
	endTime = formatTimeString(endTime)

	const validValue = /^[0-9]{2}:[0-9]{2}$/
	if (validValue.test(startTime) && validValue.test(endTime)) {
		if (new Date(`1/1/2000 ${startTime}`) > new Date(`1/1/2000 ${endTime}`)) { // if startTime bigger that endTime...
			[startTime, endTime] = [endTime, startTime] // ...swap them
		}
	}

	return (
		<StyledTimeline className={active ? 'active' : ''}>
			<div className="caption">
				<div className="caption-start">{ startTime }</div>
				<div className="caption-end">{ endTime }</div>
			</div>
			<div className="progress-bar-body">
				<div className="indicator"></div>
				<div className="line"></div>
			</div>
		</StyledTimeline>
	)
}

export default React.memo(Timeline)
