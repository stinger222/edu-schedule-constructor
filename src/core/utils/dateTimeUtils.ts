import { Cases, weekDates } from "../types/types"
import { formatNumber } from "./stringUtils"


interface IData {
  ru: {
    short: readonly string[],
    full: {
      nominative: readonly string[],
      accusative: readonly string[]
    }
  },
  en: {
    short: readonly string[],
    full: {
      nominative: readonly string[],
      accusative: readonly string[]
    }
  }
}
export class WeekDays {

  private static readonly data: IData = {
    ru: {
      short: ["Пн", "Вт", "Ср", "Чт", "Пт", "Сб", "Вс"],
      full: {
        nominative: [
          "Понедельник", "Вторник", "Среда", "Четверг", "Пятница", "Суббота", "Воскресенье"
        ],
        accusative: [
          "Понедельник", "Вторник", "Среду", "Четверг", "Пятницу", "Субботу", "Воскресенье"
        ]
      }
    },
    en: {
      short: ["M", "T", "W", "T", "F", "S", "S"],
      full: {
        nominative: [
          "Monday", "Tuesday", "Wednesday", "Thursday", "Friday", "Saturday", "Sunday"
        ],
        accusative: [
          "Monday", "Tuesday", "Wednesday", "Thursday", "Friday", "Saturday", "Sunday"
        ]
      }
    }
  }

	static getShort(lang: "ru" | "en"): ReadonlyArray<string> {
		return this.data[lang].short as ReadonlyArray<string>
	}

	static getFull(lang: "ru" | "en", caseType: Cases = Cases.Nominative, toLowerCase: boolean = false): ReadonlyArray<string> {
		const result: ReadonlyArray<string> = this.data[lang].full[caseType]
		return toLowerCase ? result.map(i => i.toLowerCase()) : result
	}
}

export const getCurrentWeekDates = (): weekDates => {
	const date = new Date()
	date.setDate(date.getDate() - (date.getDay() === 0 ? 7 : date.getDay())) // now it's current week sunday date:

	const result = Array.from({ length: 7 }, () => {
		date.setDate(date.getDate() + 1) // and now it's monday
		return formatNumber(date.getDate())
	})

	return result as weekDates
}
