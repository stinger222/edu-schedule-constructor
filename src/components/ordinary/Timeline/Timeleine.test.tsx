import { cleanup, render, screen } from "../../../core/utils/test-utils"
import Timeline from "./Timeline"


describe('Testing Timeline render with different props', () => {
	it("Renders Timeline with valid props", () => {
		render(<Timeline startTime="10:00" endTime="11:00"/>)

		expect(screen.getByText("10:00", {selector: ".caption-start"})).toBeInTheDocument()
		expect(screen.getByText("11:00", {selector: ".caption-end"})).toBeInTheDocument()
		expect(Timeline).toMatchSnapshot()
	})

	it("Renders Timeline correctly with logically incorrect props", () => {
		render(<Timeline startTime="10:00" endTime="09:00"/>)

		expect(screen.getByText("09:00", {selector: ".caption-start"})).toBeInTheDocument()
		expect(screen.getByText("10:00", {selector: ".caption-end"})).toBeInTheDocument()
		expect(Timeline).toMatchSnapshot()

		cleanup()

		render(<Timeline startTime="05:30" endTime="03:20"/>)

		expect(screen.getByText("03:20", {selector: ".caption-start"})).toBeInTheDocument()
		expect(screen.getByText("05:30", {selector: ".caption-end"})).toBeInTheDocument()
		
		render(<Timeline startTime="08:01" endTime="08:00"/>)

		expect(screen.getByText("08:00", {selector: ".caption-start"})).toBeInTheDocument()
		expect(screen.getByText("08:01", {selector: ".caption-end"})).toBeInTheDocument()
	})

	it("Renders Timeline with valid props, but values not formatted", () => {
		render(<Timeline startTime="7:3" endTime="9:5"/>)

		expect(screen.getByText("07:03", {selector: ".caption-start"})).toBeInTheDocument()
		expect(screen.getByText("09:05", {selector: ".caption-end"})).toBeInTheDocument()
		expect(Timeline).toMatchSnapshot()

		cleanup()

		render(<Timeline startTime="0:0" endTime="4:5"/>)

		expect(screen.getByText("00:00", {selector: ".caption-start"})).toBeInTheDocument()
		expect(screen.getByText("04:05", {selector: ".caption-end"})).toBeInTheDocument()
	})

	it("Renders Timeline with string props that filled with spaces", () => {
		render(<Timeline startTime="  " endTime="    "/>)

		expect(screen.getByText("??:??", {selector: ".caption-start"})).toBeInTheDocument()
		expect(screen.getByText("??:??", {selector: ".caption-end"})).toBeInTheDocument()

		cleanup()

		render(<Timeline startTime="10:00" endTime="    "/>)

		expect(screen.getByText("10:00", {selector: ".caption-start"})).toBeInTheDocument()
		expect(screen.getByText("??:??", {selector: ".caption-end"})).toBeInTheDocument()
		
		cleanup()

		render(<Timeline startTime=" " endTime="09:00"/>)

		expect(screen.getByText("??:??", {selector: ".caption-start"})).toBeInTheDocument()
		expect(screen.getByText("09:00", {selector: ".caption-end"})).toBeInTheDocument()
	})

	it("Renders Timeline with epmpty string props", () => {
		render(<Timeline startTime="" endTime=""/>)

		expect(screen.getByText("??:??", {selector: ".caption-start"})).toBeInTheDocument()
		expect(screen.getByText("??:??", {selector: ".caption-end"})).toBeInTheDocument()

		cleanup()

		render(<Timeline startTime="09:00" endTime=""/>)

		expect(screen.getByText("09:00", {selector: ".caption-start"})).toBeInTheDocument()
		expect(screen.getByText("??:??", {selector: ".caption-end"})).toBeInTheDocument()

		cleanup()

		render(<Timeline startTime="" endTime="10:00"/>)

		expect(screen.getByText("??:??", {selector: ".caption-start"})).toBeInTheDocument()
		expect(screen.getByText("10:00", {selector: ".caption-end"})).toBeInTheDocument()
	})

	it("Renders Timeline with invalid props", () => {
		render(<Timeline startTime="cum" endTime="sock"/>)
		
		expect(screen.getByText("??:??", {selector: ".caption-start"})).toBeInTheDocument()
		expect(screen.getByText("??:??", {selector: ".caption-end"})).toBeInTheDocument()
		expect(Timeline).toMatchSnapshot()

		cleanup()

		render(<Timeline startTime="08:30" endTime="sock"/>)
		
		expect(screen.getByText("08:30", {selector: ".caption-start"})).toBeInTheDocument()
		expect(screen.getByText("??:??", {selector: ".caption-end"})).toBeInTheDocument()

		cleanup()

		render(<Timeline startTime="cum" endTime="9:40"/>)
		
		expect(screen.getByText("??:??", {selector: ".caption-start"})).toBeInTheDocument()
		expect(screen.getByText("09:40", {selector: ".caption-end"})).toBeInTheDocument()
	})
})
