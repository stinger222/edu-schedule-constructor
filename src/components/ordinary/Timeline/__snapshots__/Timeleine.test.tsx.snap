// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`Testing Timeline render with different props Renders Timeline correctly with logically incorrect props 1`] = `
Object {
  "$$typeof": Symbol(react.memo),
  "compare": null,
  "type": [Function],
}
`;

exports[`Testing Timeline render with different props Renders Timeline with invalid props 1`] = `
Object {
  "$$typeof": Symbol(react.memo),
  "compare": null,
  "type": [Function],
}
`;

exports[`Testing Timeline render with different props Renders Timeline with valid props 1`] = `
Object {
  "$$typeof": Symbol(react.memo),
  "compare": null,
  "type": [Function],
}
`;

exports[`Testing Timeline render with different props Renders Timeline with valid props, but values not formatted 1`] = `
Object {
  "$$typeof": Symbol(react.memo),
  "compare": null,
  "type": [Function],
}
`;
